src/Annotation.php

```diff
@@ -63,17 +73,85 @@ public function getLabelAndUsers(array $categories): array
 
     public function getPoints(): array
     {
-        if ($this->getShape()->id === Shape::circleId()) {
-            return $this->getCirclePoints();
+        if ($this->segmentation !== null) {
+            // Handle points first (2 coordinates)
+            if (count($this->segmentation) === 2) {
+                return $this->segmentation;
+            }
+            
+            // Handle circles specially - check if it's a circle based on raw data
+            if ($this->isCircleShapeFromSegmentation()) {
+                return $this->getCirclePoints();
+            }
+            
+            return $this->segmentation;
+        }
+        
+        if ($this->bbox !== null) {
+            return $this->bboxToPoints();
+        }
+        
+        throw new \Exception("No segmentation or bbox data available");
+    }
+
+    private function isCircleShapeFromSegmentation(): bool
+    {
+        if ($this->segmentation === null) {
+            return false;
+        }
+        
+        // Tolerance for floating-point comparison
+        $tolerance = 0.001;
+        $groupedPoints = [];
+        for ($i = 0; $i < count($this->segmentation); $i += 2) {
+            $groupedPoints[] = ['x' => $this->segmentation[$i], 'y' => $this->segmentation[$i + 1]];
+        }
+
+        // Calculate the center and radius
+        $maxY = max(array_column($groupedPoints, 'y'));
+        $minY = min(array_column($groupedPoints, 'y'));
+        $maxX = max(array_column($groupedPoints, 'x'));
+        $minX = min(array_column($groupedPoints, 'x'));
+        $centerX = ($maxX + $minX) / 2;
+        $centerY = ($maxY + $minY) / 2;
+        $initialRadius = $this->euclidean_distance($groupedPoints[0]['x'], $groupedPoints[0]['y'], $centerX, $centerY);
+
+        // Check if all points are equidistant from the center
+        foreach ($groupedPoints as $point) {
+            $radius = $this->euclidean_distance($point['x'], $point['y'], $centerX, $centerY);
+            if (abs($radius - $initialRadius) > $tolerance) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private function bboxToPoints(): array
+    {
+        if ($this->bbox === null || count($this->bbox) !== 4) {
+            throw new \Exception("Invalid bbox format");
         }
-        return $this->segmentation;
+        
+        [$x, $y, $width, $height] = $this->bbox;
+        
+        return [
+            $x, $y,                    // top-left
+            $x + $width, $y,           // top-right
+            $x + $width, $y + $height, // bottom-right
+            $x, $y + $height           // bottom-left
+        ];
     }
 
     private function getGroupedPoints(): array
     {
         if ($this->groupedPoints) {
             return $this->groupedPoints;
         }
+        if ($this->segmentation === null) {
+            throw new \Exception("Cannot get grouped points without segmentation data");
+        }
```

Comment:
No need for an exception if `segmentation=[]`. Just handle empty groupedPoints in getCirclePoints() and isCircleShape().